File: examples/plugins/c/weight_truncator.c
// Example propaq native truncator plugin, implemented in C.

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define PROPAQ_TRUNCATOR_ABI_VERSION 1u

typedef struct {
    uint32_t max_weight;
} Ctx;

uint32_t propaq_truncator_abi_version(void) {
    return PROPAQ_TRUNCATOR_ABI_VERSION;
}

static uint32_t parse_max_weight(const char* config_json) {
    if (config_json == NULL) return UINT32_MAX;
    const char* key = strstr(config_json, "\"max_weight\"");
    if (key == NULL) return UINT32_MAX;
    const char* colon = strchr(key, ':');
    if (colon == NULL) return UINT32_MAX;
    return (uint32_t)strtoul(colon + 1, NULL, 10);
}

void* propaq_truncator_create(const char* config_json) {
    Ctx* ctx = (Ctx*)malloc(sizeof(Ctx));
    ctx->max_weight = parse_max_weight(config_json);
    return ctx;
}

void propaq_truncator_destroy(void* ctx) {
    free(ctx);
}

int32_t propaq_truncator_keep(void* ctx, uint32_t term_weight, double coeff_magnitude, uint32_t active_modes) {
    (void)coeff_magnitude;
    (void)active_modes;
    uint32_t max_weight = ((Ctx*)ctx)->max_weight;
    return term_weight <= max_weight ? 1 : 0;
}

int32_t propaq_truncator_keep_batch(void* ctx, const uint32_t* term_weights,
                                     const double* coeff_magnitudes, const uint32_t* active_modes,
                                     uint8_t* out_keep, size_t n) {
    (void)coeff_magnitudes;
    (void)active_modes;
    uint32_t max_weight = ((Ctx*)ctx)->max_weight;
    for (size_t i = 0; i < n; i++) {
        out_keep[i] = term_weights[i] <= max_weight ? 1 : 0;
    }
    return 0;
}

File: examples/plugins/julia/weight_truncator.jl
# Example propaq native truncator plugin, implemented in Julia.

const PROPAQ_TRUNCATOR_ABI_VERSION = UInt32(1)

const MAX_WEIGHT = UInt32(4)

Base.@ccallable function propaq_truncator_abi_version()::UInt32
    PROPAQ_TRUNCATOR_ABI_VERSION
end

Base.@ccallable function propaq_truncator_keep(ctx::Ptr{Cvoid}, term_weight::UInt32,
                                                coeff_magnitude::Cdouble, active_modes::UInt32)::Int32
    return term_weight <= MAX_WEIGHT ? Int32(1) : Int32(0)
end

Base.@ccallable function propaq_truncator_keep_batch(ctx::Ptr{Cvoid}, term_weights::Ptr{UInt32},
                                                      coeff_magnitudes::Ptr{Cdouble}, active_modes::Ptr{UInt32},
                                                      out_keep::Ptr{UInt8}, n::Csize_t)::Int32
    weights = unsafe_wrap(Array, term_weights, n)
    result = unsafe_wrap(Array, out_keep, n)
    @inbounds for i in 1:n
        result[i] = weights[i] <= MAX_WEIGHT ? UInt8(1) : UInt8(0)
    end
    return Int32(0)
end
